File: app/llm/ollama.py
import json
import requests
from app import db
from app.llm.base import LLMProvider
from app.config import (OLLAMA_HOST, OLLAMA_MODEL, OLLAMA_TIMEOUT,
                        OLLAMA_NUM_CTX, OLLAMA_NUM_PREDICT, OLLAMA_GENERATE_NUM_PREDICT)


class OllamaProvider(LLMProvider):
    def ensure_model_pulled(self) -> None:
        try:
            resp = requests.get(f"{OLLAMA_HOST}/api/tags", timeout=10)
            models = [m["name"] for m in resp.json().get("models", [])]
            model_base = OLLAMA_MODEL.split(":")[0]
            already_present = any(m.startswith(model_base) for m in models)
            if not already_present:
                print(f"Pulling model {OLLAMA_MODEL} from Ollama... (this may take a while)")
                requests.post(
                    f"{OLLAMA_HOST}/api/pull",
                    json={"name": OLLAMA_MODEL, "stream": False},
                    timeout=OLLAMA_TIMEOUT,
                )
                print(f"Model {OLLAMA_MODEL} ready.")
        except Exception as e:
            print(f"Warning: could not check/pull Ollama model: {e}")

    def classify_email_batch(self, email: dict, prompts: list) -> dict:
        if not prompts:
            return {}

        rules_text = "\n".join(
            f"{i+1}. [id:{p['id']}] {p['name']}: {p['instructions']}"
            for i, p in enumerate(prompts)
        )

        example = ", ".join(f'"{p["id"]}": false' for p in prompts[:2])
        prompt = f"""You are an email classification assistant. You will be given an email and a list of labeling rules. For each rule, decide if the label should be applied to this email.

Rules:
{rules_text}

Email:
From: {email['sender']}
Subject: {email['subject']}
Body:
{email['body'] or email['snippet']}

Respond with ONLY a JSON object where each key is the rule's [id] number and the value is true or false.
Example: {{{example}}}
No explanation, no markdown, just the JSON object."""

        try:
            response = requests.post(
                f"{OLLAMA_HOST}/api/chat",
                json={
                    "model": OLLAMA_MODEL,
                    "messages": [
                        {
                            "role": "system",
                            "content": "You are an email classification assistant. Respond only with a JSON object mapping rule IDs to true/false. No explanation, no markdown.",
                        },
                        {"role": "user", "content": prompt},
                    ],
                    "stream": False,
                    "think": False,
                    "format": "json",
                    "options": {
                        "temperature": 0,
                        "num_predict": max(50, len(prompts) * 20),
                        "num_ctx": OLLAMA_NUM_CTX,
                    },
                },
                timeout=OLLAMA_TIMEOUT,
            )
            response.raise_for_status()

            raw = response.json().get("message", {}).get("content", "").strip()

            if raw.startswith("```"):
                parts = raw.split("```")
                raw = parts[1] if len(parts) > 1 else raw
                if raw.startswith("json"):
                    raw = raw[4:]
                raw = raw.strip()

            try:
                result = json.loads(raw)
                parsed = {int(k): bool(v) for k, v in result.items()}
                db.add_log("DEBUG", f"LLM raw response: {raw}")
                db.add_log("DEBUG", f"LLM parsed: { {p['name']: parsed.get(p['id'], False) for p in prompts} }")
                return parsed
            except Exception as e:
                db.add_log("ERROR", f"LLM parse error: {e!r} | raw: {raw!r}")
                print(f"Warning: could not parse LLM batch response: {e!r} | raw: {raw!r}")
                return {p["id"]: False for p in prompts}
        except requests.exceptions.RequestException as e:
            db.add_log("ERROR", f"LLM request failed: {e!r}")
            print(f"Warning: LLM request failed: {e!r}")
            return {p["id"]: False for p in prompts}
        except Exception as e:
            db.add_log("ERROR", f"LLM unexpected error: {e!r}")
            print(f"Warning: LLM unexpected error: {e!r}")
            return {p["id"]: False for p in prompts}

    def _build_generate_request(self, description: str) -> dict:
        system_prompt = (
            "You write precise email filter rules for an AI classifier. "
            "The rule must specify:\n"
            "- Exact positive signals (what SHOULD trigger the label)\n"
            "- What should NOT trigger it (to prevent false positives)\n"
            "- Specific sender domains, subject keywords, or body patterns where applicable\n"
            "Output only the rule text. No preamble, no quotes, no explanation."
        )
        user_prompt = (
            f'A user wants to automatically label certain emails. They described:\n\n"{description}"\n\n'
            "Write a precise classifier instruction (2-5 sentences). Be concrete:\n"
            "- State what specific content, sender patterns, or subject keywords indicate a match\n"
            "- State what distinguishes these emails from similar-but-different ones (to avoid false positives)\n"
            "- If the description is generic (e.g. \"promotional emails\"), name specific signals: "
            "unsubscribe links, \"% off\", bulk sender headers, marketing domains, etc.\n\n"
            "Respond with ONLY the instruction text."
        )
        return {
            "model": OLLAMA_MODEL,
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt},
            ],
            "options": {
                "temperature": 0.7,
                "num_predict": OLLAMA_GENERATE_NUM_PREDICT,
                "num_ctx": OLLAMA_NUM_CTX,
            },
        }

    def _filter_think_chunks(self, buffer: str, in_think: bool, chunk: str):
        """Append chunk to buffer, flush safe content, return (events, new_buffer, in_think).
        Events are (type, text) tuples where type is 'think' or 'content'."""
        buffer += chunk
        events = []
        while True:
            tag = "</think>" if in_think else "<think>"
            idx = buffer.find(tag)
            if idx == -1:
                # No complete tag — keep up to len(tag)-1 chars buffered
                safe = max(0, len(buffer) - (len(tag) - 1))
                if safe > 0:
                    events.append(("think" if in_think else "content", buffer[:safe]))
                    buffer = buffer[safe:]
                break
            else:
                if idx > 0:
                    events.append(("think" if in_think else "content", buffer[:idx]))
                buffer = buffer[idx + len(tag):]
                in_think = not in_think
        return events, buffer, in_think

    def stream_generate_prompt_instruction(self, description: str):
        """Generator that yields {"type": "think"|"content", "text": str} dicts."""
        payload = self._build_generate_request(description)
        payload["stream"] = True
        response = requests.post(
            f"{OLLAMA_HOST}/api/chat",
            json=payload,
            stream=True,
            timeout=OLLAMA_TIMEOUT,
        )
        response.raise_for_status()
        in_think = False
        buffer = ""
        for line in response.iter_lines():
            if not line:
                continue
            try:
                data = json.loads(line)
            except Exception:
                continue
            token = data.get("message", {}).get("content", "")
            if not token:
                continue
            events, buffer, in_think = self._filter_think_chunks(buffer, in_think, token)
            for evt_type, evt_text in events:
                if evt_text:
                    yield {"type": evt_type, "text": evt_text}
        # Flush remaining buffer
        if buffer:
            yield {"type": "think" if in_think else "content", "text": buffer}

    def generate_prompt_instruction(self, description: str) -> str:
        payload = self._build_generate_request(description)
        payload["stream"] = False
        response = requests.post(
            f"{OLLAMA_HOST}/api/chat",
            json=payload,
            timeout=OLLAMA_TIMEOUT,
        )
        response.raise_for_status()
        content = response.json().get("message", {}).get("content", "").strip()
        # Strip think blocks in case model includes them despite stream=False
        import re
        content = re.sub(r'<think>.*?</think>', '', content, flags=re.DOTALL).strip()
        return content

File: llm/ollama.go
package llm

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"regexp"
	"strings"
	"time"

	"github.com/sloccy/ollamail/db"
)

var fenceRe = regexp.MustCompile(`(?s)^` + "```" + `(?:json)?\s*|\s*` + "```" + `$`)

type Client struct {
	host       string
	model      string
	numCtx     int
	timeout    time.Duration
	httpClient *http.Client
}

func NewClient(host, model string, numCtx int, timeout time.Duration) *Client {
	return &Client{
		host:       strings.TrimRight(host, "/"),
		model:      model,
		numCtx:     numCtx,
		timeout:    timeout,
		httpClient: &http.Client{Timeout: timeout},
	}
}

func (c *Client) Model() string { return c.model }

// ============================================================
// Model management
// ============================================================

func (c *Client) EnsureModelPulled(store *db.Store) error {
	ctx := context.Background()
	exists, err := c.modelExists(ctx)
	if err != nil {
		store.Log("WARNING", fmt.Sprintf("Could not check Ollama model: %v", err))
		return err
	}
	if !exists {
		store.Log("INFO", fmt.Sprintf("Pulling model %s from Ollama... (this may take a while)", c.model))
		if err := c.pullModel(ctx); err != nil {
			store.Log("WARNING", fmt.Sprintf("Could not pull Ollama model: %v", err))
			return err
		}
		store.Log("INFO", fmt.Sprintf("Model %s ready.", c.model))
	}
	return nil
}

func (c *Client) modelExists(ctx context.Context) (bool, error) {
	body, err := json.Marshal(map[string]string{"model": c.model})
	if err != nil {
		return false, err
	}
	req, _ := http.NewRequestWithContext(ctx, http.MethodPost, c.host+"/api/show", bytes.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return false, err
	}
	defer func() { _ = resp.Body.Close() }()
	return resp.StatusCode == http.StatusOK, nil
}

func (c *Client) pullModel(ctx context.Context) error {
	body, err := json.Marshal(map[string]any{"model": c.model, "stream": true})
	if err != nil {
		return err
	}
	req, _ := http.NewRequestWithContext(ctx, http.MethodPost, c.host+"/api/pull", bytes.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	// Use unlimited timeout client for pull
	resp, err := (&http.Client{}).Do(req)
	if err != nil {
		return err
	}
	defer func() { _ = resp.Body.Close() }()
	// Drain streaming response
	_, _ = io.Copy(io.Discard, resp.Body)
	return nil
}

// ============================================================
// Classification
// ============================================================

type Email struct {
	Sender  string
	Subject string
	Body    string
	Snippet string
}

type Prompt struct {
	ID           int64
	Name         string
	Instructions string
}

// Error is returned when the LLM fails to produce a usable response.
type Error struct{ Msg string }

func (e *Error) Error() string { return e.Msg }

func (c *Client) ClassifyEmailBatch(ctx context.Context, store *db.Store, email Email, prompts []Prompt) (map[int64]bool, string, string, error) {
	if len(prompts) == 0 {
		return nil, "", "", nil
	}

	body := buildBody(email, prompts)
	numPredict := 50
	if n := len(prompts) * 20; n > numPredict {
		numPredict = n
	}

	payload := map[string]any{
		"model": c.model,
		"messages": []map[string]string{
			{
				"role":    "system",
				"content": "You are an email classification assistant. Respond only with a JSON object mapping rule numbers to true/false. No explanation, no markdown.",
			},
			{"role": "user", "content": body},
		},
		"think":  false,
		"format": "json",
		"stream": false,
		"options": map[string]any{
			"temperature": 0,
			"num_predict": numPredict,
			"num_ctx":     c.numCtx,
		},
	}

	requestBytes, err := json.Marshal(payload)
	if err != nil {
		requestBytes = []byte("{}")
	}
	requestJSON := string(requestBytes)

	subject := email.Subject
	if len(subject) > 60 {
		subject = subject[:60]
	}
	store.Log("INFO", fmt.Sprintf("LLM classifying '%s' against %d rule(s)", subject, len(prompts)))

	raw, err := c.doChat(ctx, payload)
	if err != nil {
		store.Log("ERROR", fmt.Sprintf("LLM request failed: %v", err))
		return nil, requestJSON, "", &Error{Msg: fmt.Sprintf("LLM request failed: %v", err)}
	}

	store.Log("INFO", fmt.Sprintf("LLM classify response: content=%d chars", len(raw)))
	if len(raw) > 0 {
		preview := raw
		if len(preview) > 500 {
			preview = preview[:500]
		}
		store.Log("INFO", "LLM raw content: "+preview)
	}

	rawResponse := raw
	raw = strings.TrimSpace(fenceRe.ReplaceAllString(raw, ""))

	var result map[string]any
	if err := json.Unmarshal([]byte(raw), &result); err != nil {
		store.Log("ERROR", fmt.Sprintf("LLM parse error: %v | raw: %s", err, raw))
		return nil, requestJSON, rawResponse, &Error{Msg: fmt.Sprintf("LLM parse error: %v", err)}
	}

	parsed := make(map[int64]bool, len(prompts))
	for k, v := range result {
		var idx int
		if _, err := fmt.Sscanf(k, "%d", &idx); err != nil {
			continue
		}
		idx-- // 1-based to 0-based
		if idx >= 0 && idx < len(prompts) {
			b, _ := v.(bool)
			parsed[prompts[idx].ID] = b
		}
	}
	return parsed, requestJSON, rawResponse, nil
}

func buildBody(email Email, prompts []Prompt) string {
	var sb strings.Builder
	for i, p := range prompts {
		fmt.Fprintf(&sb, "%d. %s: %s\n", i+1, p.Name, p.Instructions)
	}
	rulesText := sb.String()

	exampleParts := make([]string, min(2, len(prompts)))
	for i := range exampleParts {
		exampleParts[i] = fmt.Sprintf(`"%d": false`, i+1)
	}

	body := email.Body
	if body == "" {
		body = email.Snippet
	}

	return fmt.Sprintf(`You are an email classification assistant. You will be given an email and a list of labeling rules. For each rule, decide if the label should be applied to this email.

Rules:
%s
Email:
From: %s
Subject: %s
Body:
%s

Respond with ONLY a JSON object where each key is the rule's number (1, 2, 3...) and the value is true or false.
Example: {%s}
No explanation, no markdown, just the JSON object.`,
		rulesText, email.Sender, email.Subject, body,
		strings.Join(exampleParts, ", "))
}

// ============================================================
// Streaming prompt generation
// ============================================================

// StreamChunk is yielded by StreamGeneratePromptInstruction.
type StreamChunk struct {
	Text string
	Err  error
}

func (c *Client) StreamGeneratePromptInstruction(ctx context.Context, description string) <-chan StreamChunk {
	ch := make(chan StreamChunk, 16)
	go func() {
		defer close(ch)
		if err := c.streamGenerate(ctx, description, ch); err != nil {
			select {
			case ch <- StreamChunk{Err: err}:
			case <-ctx.Done():
			}
		}
	}()
	return ch
}

func (c *Client) streamGenerate(ctx context.Context, description string, ch chan<- StreamChunk) error {
	payload := map[string]any{
		"model": c.model,
		"messages": []map[string]string{
			{
				"role":    "system",
				"content": "You write email filter rules for an AI classifier. Output only the rule text. No preamble, no drafts, no self-critique, no quotes, no explanation.",
			},
			{
				"role": "user",
				"content": fmt.Sprintf(
					"Write a 2-4 sentence classifier instruction for emails matching: %q\n\n"+
						"The instruction must describe: what the email is about, its purpose/intent, "+
						"and what distinguishes it from similar-but-non-matching emails. "+
						"Do not use keywords or sender addresses as criteria — focus on meaning and context.\n\n"+
						"Output ONLY the instruction text.",
					description),
			},
		},
		"stream": true,
		"options": map[string]any{
			"temperature": 0.7,
			"num_predict": 2048,
			"num_ctx":     c.numCtx,
		},
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.host+"/api/chat", bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer func() { _ = resp.Body.Close() }()

	scanner := bufio.NewScanner(resp.Body)
	for scanner.Scan() {
		line := scanner.Bytes()
		if len(line) == 0 {
			continue
		}
		var chunk struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
			Done bool `json:"done"`
		}
		if err := json.Unmarshal(line, &chunk); err != nil {
			continue
		}
		if chunk.Message.Content != "" {
			select {
			case ch <- StreamChunk{Text: chunk.Message.Content}:
			case <-ctx.Done():
				return ctx.Err()
			}
		}
		if chunk.Done {
			break
		}
	}
	slog.Info("stream_generate finished")
	return scanner.Err()
}

// ============================================================
// Prompt Improvement
// ============================================================

// ChatMessage is a single turn in a conversation.
type ChatMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// ImproveRequest carries everything needed to generate an improved prompt instruction.
type ImproveRequest struct {
	PromptName           string
	LabelName            string
	OriginalInstructions string
	TriggerKind          string // "false_negative" | "false_positive"
	EmailSubject         string
	EmailSender          string
	EmailBody            string
	PriorConversation    []ChatMessage
	UserComment          string
}

const improveSystemPrompt = `You are a careful editor of email-classification rules. You are given one existing rule (its name, target Gmail label, and current instructions) and one concrete email that the rule handled incorrectly. Your job is to rewrite the instructions so that the same rule would have handled this email correctly, without damaging its behavior on emails it currently classifies correctly.

CRITICAL OUTPUT REQUIREMENT: Your entire response must be ONLY the rewritten rule instructions — nothing else. No preamble, no explanation, no "Here is the updated rule:", no quoting of the email, no markdown formatting, no commentary. Think as long as you need internally, but the only thing you output is the new instructions text itself.

Rules for rewriting:
1. Preserve the rule's original intent. Do not widen scope beyond what the name and label imply. Do not turn a narrow rule into a catch-all.
2. Never use the specific sender address, subject line, or body phrases of the example email as matching criteria. The example is an illustration, not a fingerprint. Match on meaning, purpose, and context.
3. If trigger_kind is false_negative: explain what category of email was missed and add language that would match it. If trigger_kind is false_positive: add exclusions or clarify the scope so emails like this one are no longer matched.
4. Keep the output 2-6 sentences. Plain prose. No bullet lists, no code blocks, no markdown headings.
5. If the user comments on your suggestion, treat the comment as authoritative feedback and produce another revision that addresses it while still obeying rules 1-4.

Remember: output ONLY the rewritten instructions text. No other text whatsoever.`

// ImprovePromptInstructions calls the LLM to rewrite a prompt's instructions based on
// a misclassification example. It returns the revised text, the full conversation (for
// subsequent iterations), and any error.
func (c *Client) ImprovePromptInstructions(ctx context.Context, req ImproveRequest) (string, []ChatMessage, error) {
	messages := []map[string]string{{"role": "system", "content": improveSystemPrompt}}

	if len(req.PriorConversation) > 0 {
		for _, m := range req.PriorConversation {
			messages = append(messages, map[string]string{"role": m.Role, "content": m.Content})
		}
		// Append latest user comment
		messages = append(messages, map[string]string{
			"role":    "user",
			"content": req.UserComment,
		})
	} else {
		userMsg := fmt.Sprintf(
			"RULE NAME: %s\nTARGET LABEL: %s\nTRIGGER: %s\n\nCURRENT INSTRUCTIONS:\n%s\n\nMISHANDLED EMAIL:\nFrom: %s\nSubject: %s\nBody:\n%s\n\nRewrite the instructions per the system rules.",
			req.PromptName, req.LabelName, req.TriggerKind,
			req.OriginalInstructions,
			req.EmailSender, req.EmailSubject, req.EmailBody,
		)
		messages = append(messages, map[string]string{"role": "user", "content": userMsg})
	}

	payload := map[string]any{
		"model":    c.model,
		"messages": messages,
		"stream":   false,
		"options": map[string]any{
			"temperature": 0.4,
			"num_predict": 16384,
			"num_ctx":     c.numCtx,
		},
	}

	raw, err := c.doChat(ctx, payload)
	if err != nil {
		return "", nil, err
	}
	suggestion := strings.TrimSpace(raw)

	// Build updated conversation for storage
	var conv []ChatMessage
	conv = append(conv, req.PriorConversation...)
	if len(req.PriorConversation) > 0 {
		conv = append(conv, ChatMessage{Role: "user", Content: req.UserComment})
	} else {
		conv = append(conv, ChatMessage{Role: "user", Content: fmt.Sprintf(
			"RULE NAME: %s\nTARGET LABEL: %s\nTRIGGER: %s\n\nCURRENT INSTRUCTIONS:\n%s\n\nMISHANDLED EMAIL:\nFrom: %s\nSubject: %s\nBody:\n%s\n\nRewrite the instructions per the system rules.",
			req.PromptName, req.LabelName, req.TriggerKind,
			req.OriginalInstructions,
			req.EmailSender, req.EmailSubject, req.EmailBody,
		)})
	}
	conv = append(conv, ChatMessage{Role: "assistant", Content: suggestion})

	return suggestion, conv, nil
}

// ============================================================
// Internal
// ============================================================

func (c *Client) doChat(ctx context.Context, payload map[string]any) (string, error) {
	body, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.host+"/api/chat", bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return "", err
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("ollama %d: %s", resp.StatusCode, string(b))
	}

	var result struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", err
	}
	return result.Message.Content, nil
}
